### PR TITLE
Make Float8 types serializeable

### LINKED ISSUE #114634: float8 and bits tensors cannot be serialized
```python
import torch
t=torch.empty(10, dtype=torch.float8_e5m2)
torch.save(t, "/tmp/t")
```
results in
```
File ~/.pyenv/versions/3.11.5/lib/python3.11/site-packages/torch/storage.py:908, in TypedStorage._pickle_storage_type(self)
    906     return _dtype_to_storage_type_map()[self.dtype]
    907 except KeyError as e:
--> 908     raise KeyError(f'dtype {self.dtype} is not recognized') from e

KeyError: 'dtype torch.float8_e5m2 is not recognized'
```
the comment next to `_dtype_to_storage_type_map` says that no new types should be added to that map and that new tensors should be serialized with untyped storage, and yet I didn't see how to take untyped storage path. 

cc @ezyang @gchanan @zou3519 @kadeng @mruberry @mikaylagawarecki

### PR DESCRIPTION
**Pull request by malfet**
By finally breaking FC promise on new dtypes by serializing untyped
storage and tensor dtypes

- Add `_rebuild_tensor_v3` that takes an extra dtype argument
- In `Tensor.__reduce_ex__` serialize tensor using untyped storage for
  v3_dtypes (which are at the moment limited to float8 dtypes)

Test plan: `python -c "import torch;x=torch.arange(10).to(dtype=torch.float8_e4m3fn);torch.save(x, 'pt.pt');print(torch.load('pt.pt'))"`

Fixes https://github.com/pytorch/pytorch/issues/114634

### CODE REVIEW

File: torch/_tensor.py [hunk 1]
```diff
@@ -394,25 +394,36 @@ def _reduce_ex_internal(self, proto):
             )
             return (torch._utils._rebuild_wrapper_subclass, arg_wrapper_subclass)
         else:
-            # TODO: Once we decide to break serialization FC, no longer
-            # need to wrap with TypedStorage
-            args = (
-                torch.storage.TypedStorage(
+            v3_dtypes = [torch.float8_e5m2, torch.float8_e4m3fn]
```
Comment (ngimel) [inline torch/_tensor.py:397]: also `uz` dtypes?

Comment (ngimel) [reply torch/_tensor.py:397]: Probably makes sense to serialize all tensors whose dtype is not in the dict like this.

**Review by ngimel — APPROVED** [review]

File: torch/_tensor.py (hunk 1 above)
Comment (malfet) [reply torch/_tensor.py]: Hmm, so do the inverse. Makes sense, will do in follow up PR (and probably past the branch cut, as it feels a bit risky)

**Pull request closed without merging**